ecs-cli/modules/cli/local/network/network.go

```diff
@@ -0,0 +1,217 @@
+func setupLocalNetwork(docker idempotentNetworkCreator) {
+	if localNetworkExists(docker) {
+		logrus.Infof("The network %s already exists", EcsLocalNetworkName)
+		return
+	}
+	createLocalNetwork(docker)
+}
+
+// localNetworkExists returns true if the local network already exists, false otherwise.
+func localNetworkExists(docker idempotentNetworkCreator) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), dockerTimeout)
+	defer cancel()
+
+	_, err := docker.NetworkInspect(ctx, EcsLocalNetworkName, types.NetworkInspectOptions{})
+	if err != nil {
+		if client.IsErrNotFound(err) {
+			return false
+		}
+		// Unexpected error while inspecting docker networks, we want to crash the app.
+		logrus.Fatalf("Failed to inspect docker network %s due to %v", EcsLocalNetworkName, err)
+	}
+	return true
+}
+
+// createLocalNetwork creates the local network.
+func createLocalNetwork(docker idempotentNetworkCreator) {
+	ctx, cancel := context.WithTimeout(context.Background(), dockerTimeout)
+	defer cancel()
+
+	resp, err := docker.NetworkCreate(ctx, EcsLocalNetworkName, types.NetworkCreate{
+		IPAM: &network.IPAM{
+			Config: []network.IPAMConfig{
+				{
+					Subnet: ecsLocalNetworkSubnet,
+				},
+			},
+		},
+	})
+	if err != nil {
+		logrus.Fatalf("Failed to create network %s with subnet %s due to %v", EcsLocalNetworkName, ecsLocalNetworkSubnet, err)
+	}
+	logrus.Infof("Created network %s with ID %s", EcsLocalNetworkName, resp.ID)
+}
+
+// setupLocalEndpointsContainer creates the Local Endpoints container if one doesn't exist already.
+func setupLocalEndpointsContainer(docker idempotentContainerStarter) {
+	containerID := createLocalEndpointsContainer(docker)
+	startContainer(docker, containerID)
+}
+
+// createLocalEndpointsContainer returns the ID of the newly created container.
+// If the container already exists, returns the ID of the existing container.
+func createLocalEndpointsContainer(docker idempotentContainerStarter) string {
+	ctx, cancel := context.WithTimeout(context.Background(), dockerTimeout)
+	defer cancel()
+
+	// See First Scenario in https://aws.amazon.com/blogs/compute/a-guide-to-locally-testing-containers-with-amazon-ecs-local-endpoints-and-docker-compose/
+	// for an explanation of these fields.
+	resp, err := docker.ContainerCreate(ctx,
+		&container.Config{
+			Image: localEndpointsImageName,
+			Env: []string{
+				"AWS_PROFILE=default",
+				"HOME=/home",
+			},
+		},
+		&container.HostConfig{
+			Binds: []string{
+				"/var/run:/var/run",
+				fmt.Sprintf("%s/.aws/:/home/.aws/", os.Getenv("HOME")),
+			},
+		},
+		&network.NetworkingConfig{
+			EndpointsConfig: map[string]*network.EndpointSettings{
+				EcsLocalNetworkName: {
+					NetworkID: EcsLocalNetworkName,
+					IPAMConfig: &network.EndpointIPAMConfig{
+						IPv4Address: localEndpointsContainerIpAddr,
+					},
+				},
+			},
+		},
+		localEndpointsContainerName,
+	)
+	if err != nil {
+		if strings.Contains(err.Error(), "Conflict") {
```

Comment:
Let me make sure I understand this correctly. So what happens in the future when someone updates the Docker SDK and Docker changes their error message to not include the string "Conflict"?

The CLI command will just fail and then print the Docker error message right?
So as long as Docker keeps its error messages descriptive, the user will know what happened.

Reply:
This is really annoying, I didn't want to use string matching in the error but there is no other alternative because of how Docker handles errors from the daemon (see https://github.com/moby/moby/blob/28d7dba41d0c0d9c7f0dafcc79d3c59f2b3f5dc3/client/request.go#L236)

So we don't have any way of validating if the error is a conflict, like [IsConflict](https://github.com/moby/moby/blob/master/errdefs/is.go#L44), anymore :(

As you said, if they remove "Conflict" from the error message then we will crash and log the new error. Hopefully, our integ tests can catch these sort of issues before a customer contacts us.

Reply:
> if they remove "Conflict" from the error message then we will crash and log the new error

Cool. I just wanted to verify that. While this is not great code, I understand that there is no other choice for now, so I'm fine with it.